### ISSUE TITLE
A CrystallographicMap class

### ISSUE THREAD

**Issue by hakonanes** [pyxem/orix#48]
I'm motivated to create a `CrystallographicMap` class with `Rotation` objects and quality metric/property maps to couple these to EBSD pattern analysis in [KikuchiPy](https://github.com/kikuchipy/kikuchipy).

See start of discussion on this class' implementation here: https://github.com/pyxem/orix/issues/47#issuecomment-593078559.

It should have existing functionality available in pyXem (https://github.com/pyxem/pyxem/blob/master/pyxem/signals/crystallographic_map.py).

Disclaimer: I haven't used orix before, and don't know all the `CrystallographicMap` class' potential use cases in orix, or things that need to change (if any) in the package overall (or how it affects other packages depending on it).

Briefly, I would like something similar to the following

```python
>>> import orix
>>> import matplotlib.pyplot as plt
>>> data = orix.io.load_h5ebsd('/my/file.h5')
>>> data.__class__
orix.CrystallographicMap
>>> data.orientations
# The Rotation object
>>> data.prop
# Listing all available property maps
>>> plt.imshow(data.prop['image_quality'])  # Gives me a nice 2D image quality map
```

A new property can easily be added with `data.prop['my_new_property']`.

A natural source of inspiration for this class is MTEX' EBSD class (https://github.com/mtex-toolbox/mtex/tree/develop/EBSDAnalysis/%40EBSD).

**Comment by pc494** [comment]
This seems like an excellent, light-weight addition to the code base. In terms of internal structure might I suggest using `(N,m,l)` numpy arrays, where `N` is the number of metrics, and `m*l` is the data shape.

Then, with a short lookup table something like `data.prop` could return the relevant sub-array. I would keep the `Rotation` object separate from these arrays as they carry a lot of extra code.

The obvious (simplest) use case is perhaps combining "Rotation" and its clustering labels as in the demo notebooks?

**Comment by hakonanes** [comment]
Sorry Phillip, but I don't understand your suggestion for an internal structure. You mean *not* use a dictionary of `N` numpy arrays of shape `(l, m)`? But have *one* numpy array with all properties?

I agree with keeping `Rotation` objects separate from this property dictionary (or numpy array). What should we call the class attribute? `CrystallographicMap.rotations`? Then all `Rotation` methods can be called on `data.rotations.some_rotation_method()`.

Where should this class go, and how should it be exposed in the API? Should I add the class `CrystallographicMap` in its own directory, `orix/orix/crystallographic_map` to either a `crystallographic_map.py` or to the `__init__.py`? Or simply call it `Map` (which might be too general)? I myself would want to at import it with something simple, like `orix.CrystallographicMap(my_stuff)`.

I'm asking all these questions because the package, to my eyes, is tidy and nicely structured, and I don't to want to change that.

**Comment by dnjohnstone** [comment]
I must say I'm not quite sure what @pc494 is intending to say - but I think he is suggesting the one numpy array approach with a dictionary to convert from a metric defining string to a number to index the array?

 In response to your other points:

> I agree with keeping `Rotation` objects separate from this property dictionary (or numpy array). What should we call the class attribute? `CrystallographicMap.rotations`? Then all `Rotation` methods can be called on `data.rotations.some_rotation_method()`.

Yes I think that's what I have in mind. I think there will be lots of methods to the CrystallographicMap class - e.g. get_orientation_map(), that will mean that in the long run using the Rotation class methods directly will be a relatively "advanced" use case.

> Where should this class go, and how should it be exposed in the API? Should I add the class `CrystallographicMap` in its own directory, `orix/orix/crystallographic_map` to either a `crystallographic_map.py` or to the `__init__.py`? Or simply call it `Map` (which might be too general)? I myself would want to at import it with something simple, like `orix.CrystallographicMap(my_stuff)`.

Map is definitely too general. But the right answer isn't immediately obvious to me. I think we need to think a little ahead to other things that might eventually go in to orix that are somehow related to a CrystallographicMap - which is really a "class for orientation data associated with many crystals and possibly many types of crystals"...

Some assorted thoughts are:

1. I think we're all imagining two-dimensional maps of <x, y | phase, alpha, beta, gamma> dimensions right now, but we should also support line scans <x | phase, alpha, beta, gamma> and (perhaps not immediately) three dimensional maps <x, y, z | phase, alpha, beta, gamma> - we should probably decide now whether we'd want that to be 3 different classes, or 1 class. I think 3.

2. Shouldn't we be using Orientation() objects rather than Rotation() objects so that the crystal symmetry goes with it that way?

3. I feel like we need to think about how other obvious additions e.g. pole figures, and orientation distribution functions are going to play with this bit of the code. But I think it might be too difficult to predict a good architecture for all that from where we are right now.

So based on that I think we should put this in orix/crystal_maps/crystal_map2d.py

What do others think?

**Comment by pc494** [comment]
Just to address point (1), I was actually thinking `<phase,metric,orientation| x y z>`, which is probably easier to generalize to 1/2/3D (and more in-keeping with the code style at present). Then 1 class seems better.

**Comment by hakonanes** [comment]
`CrystalMap` sounds good, and placing it under `orix/crystal_maps/crystal_map2d.py`, and calling it via `orix.crystal_maps.CrystalMap`.

And yes, the `Orientation` class is of course best. I was just a bit confused with this since the `loadang` and `loadctf` return `Rotation` objects. Should they instead return `Orientation`?

We could also store phase, orientations and properties in one long line internally, but have X, Y and Z coordinates for them so we know where they are in a line/map/cube? For this I don't think we would need to generalize to 1D and 3D? Is this what you suggest, @pc494? And could you point me to an example in the code where the style you mention is used?

**Comment by dnjohnstone** [comment]
> And yes, the `Orientation` class is of course best. I was just a bit confused with this since the `loadang` and `loadctf` return `Rotation` objects. Should they instead return `Orientation`?

It might be that this is a pain if the symmetry is not stored with the orientation data in .ang / .ctf file. You'd probably want to be able to load it and look at it before setting the symmetry. I think this is basically what you do in MTEX too if I recall correctly. So maybe it should generally load in to a Rotation class somewhere and then be converted to Orientation (as we currently do).

> We could also store phase, orientations and properties in one long line internally, but have X, Y and Z coordinates for them so we know where they are in a line/map/cube? For this I don't think we would need to generalize to 1D and 3D? Is this what you suggest, @pc494? And could you point me to an example in the code where the style you mention is used?

I think this is what @pc494 is basically suggesting - and as usual when it comes to this stuff he's probably right. In this case I guess we can still put it in orix/crystal_map/crystal_map.py - does anyone have a better suggestion than this.

The code in pyxem is really very similar to the code in MTEX (as we acknowledge this is very much our inspiration) so the bit of code that this should be somehow similar to is the MTEX EBSD class https://github.com/mtex-toolbox/mtex/blob/develop/EBSDAnalysis/%40EBSD/EBSD.m

Obviously we're going with a more general name because we're imagining not only EBSD data but more general mapping data.

I do still wonder whether we should be separating different spatial dimensions though - only because there could be some things e.g. analysis to get a Nye tensor that might need different implementations for different dimensionalities. But  this may be something to muddle out in future.

**Comment by hakonanes** [comment]
OK, great. And thanks for the MTEX reference. I think I have an understanding of how to set things up now.

**Comment by pc494** [comment]
This sounds really good. Some things that sprung to mind:

- I eventually decided that using the `Scalar` class for things like `image_quality` is the way to go. 
- I think at this time we should make `CrystalMap` (ie not `CrystalMap2D`)and then at the point we write something that's dimensionally dependent (Kernal Average Misorientation springs to mind) figure out the best approach. 
- PR early because life is just easier that way.

Looking forward to seeing to seeing the results of this.

**Comment by hakonanes** [comment]
@pc494 Took you up on the PR early comment. It's extremely WIP, but this way I won't code too much into a corner...

**Comment by pc494** [comment]
Am I correct in thinking this can be closed now? (@hakonanes)

**Comment by hakonanes** [comment]
Closed due to #49!
